File: src/main/java/io/github/malczuuu/iemu/domain/FirmwareService.java
package io.github.malczuuu.iemu.domain;

import io.github.malczuuu.iemu.infrastructure.lwm2m.FirmwareUpdateResult;
import io.github.malczuuu.iemu.infrastructure.lwm2m.FirmwareUpdateState;
import java.util.function.Consumer;
import java.util.function.IntConsumer;

public interface FirmwareService {

  FirmwareDTO getFirmware();

  void changeFirmware(FirmwareDTO firmware);

  void executeFirmwareUpdate();

  void subscribeOnFileChange(Consumer<byte[]> consumer);

  void subscribeOnPackageUriChange(Consumer<String> consumer);

  void subscribeOnStateChange(Consumer<FirmwareUpdateState> consumer);

  void subscribeOnResultChange(Consumer<FirmwareUpdateResult> consumer);

  void subscribeOnPackageVersionChange(Consumer<String> consumer);

  void subscribeOnProgressChange(IntConsumer consumer);
}

File: src/main/kotlin/io/github/malczuuu/iemu/domain/FirmwareService.kt
package io.github.malczuuu.iemu.domain

import io.github.malczuuu.iemu.infra.lwm2m.FirmwareUpdateResult
import io.github.malczuuu.iemu.infra.lwm2m.FirmwareUpdateState

interface FirmwareService {

  fun start()

  fun getFirmware(): Firmware

  fun changeFirmware(update: FirmwareUpdate)

  fun executeFirmwareUpdate()

  fun subscribeOnFileChange(consumer: (ByteArray) -> Unit): Boolean

  fun subscribeOnPackageUriChange(consumer: (String?) -> Unit): Boolean

  fun subscribeOnStateChange(consumer: (FirmwareUpdateState) -> Unit): Boolean

  fun subscribeOnResultChange(consumer: (FirmwareUpdateResult) -> Unit): Boolean

  fun subscribeOnPackageVersionChange(consumer: (String?) -> Unit): Boolean

  fun subscribeOnProgressChange(consumer: (Int) -> Unit): Boolean
}
